src/Runner.Listener/JobDispatcher.cs

```diff
@@ -546,13 +546,27 @@ private async Task RunAsync(Pipelines.AgentJobRequestMessage message, string orc
                                     Trace.Info($"Return code {returnCode} indicate worker encounter an unhandled exception or app crash, attach worker stdout/stderr to JobRequest result.");
 
                                     var jobServer = await InitializeJobServerAsync(systemConnection);
-                                    await LogWorkerProcessUnhandledException(jobServer, message, detailInfo);
-
-                                    // Go ahead to finish the job with result 'Failed' if the STDERR from worker is System.IO.IOException, since it typically means we are running out of disk space.
-                                    if (detailInfo.Contains(typeof(System.IO.IOException).ToString(), StringComparison.OrdinalIgnoreCase))
+                                    var unhandledExceptionIssue = new Issue() { Type = IssueType.Error, Message = detailInfo };
+                                    unhandledExceptionIssue.Data[Constants.Runner.InternalTelemetryIssueDataKey] = Constants.Runner.WorkerCrash;
+                                    switch (jobServer)
                                     {
-                                        Trace.Info($"Finish job with result 'Failed' due to IOException.");
-                                        await ForceFailJob(jobServer, message, detailInfo);
+                                        case IJobServer js:
+                                        {
+                                            await LogWorkerProcessUnhandledException(js, message, unhandledExceptionIssue);
+                                            // Go ahead to finish the job with result 'Failed' if the STDERR from worker is System.IO.IOException, since it typically means we are running out of disk space.
+                                            if (detailInfo.Contains(typeof(System.IO.IOException).ToString(), StringComparison.OrdinalIgnoreCase))
+                                            {
+                                                Trace.Info($"Finish job with result 'Failed' due to IOException.");
+                                                await ForceFailJob(js, message);
+                                            }
+
+                                            break;
+                                        }
+                                        case IRunServer rs:
+                                            await ForceFailJob(rs, message, unhandledExceptionIssue);
```

Comment:
only do this for `if (detailInfo.Contains(typeof(System.IO.IOException).ToString(), StringComparison.OrdinalIgnoreCase))` to match existing behavior?

Reply:
existing one already does failing job, inside `LogWorkerProcessUnhandledException`:

```
if (message.Variables.TryGetValue("DistributedTask.MarkJobAsFailedOnWorkerCrash", out var markJobAsFailedOnWorkerCrash) &&
                    StringUtil.ConvertToBoolean(markJobAsFailedOnWorkerCrash?.Value))
                {
                    Trace.Info("Mark the job as failed since the worker crashed");
                    jobRecord.Result = TaskResult.Failed;
                    // mark the job as completed so service will pickup the result
                    jobRecord.State = TimelineRecordState.Completed;
                }
```

Reply:
that ff is ON, so I think we still match with existing behavior overall